snowfall/training/ctc_graph.py

```diff
@@ -0,0 +1,159 @@
+    Returns:
+      Returns an FSA that converts repeated symbols to a single symbol.
+    '''
+    num_states = len(phones)
+    final_state = num_states
+    rules = ''
+    for i in range(num_states):
+        for j in range(num_states):
+            if i == j:
+                rules += f'{i} {i} {phones[i]} 0 0.0\n'
+            else:
+                rules += f'{i} {j} {phones[j]} {phones[j]} 0.0\n'
+        rules += f'{i} {final_state} -1 -1 0.0\n'
+    rules += f'{final_state}'
+    ans = k2.Fsa.from_str(rules)
+    return ans
+
+
+def compose(a_fsa: k2.Fsa, b_fsa: k2.Fsa) -> k2.Fsa:
+    '''Compose two graphs.
+
+    Args:
+      a_fsa:
+        The FSA on the left hand side.
+      b_fsa:
+        The FSA on the right hand side.
+
+    Returns:
+      The composition result. It will have:
+    '''
+    if not hasattr(a_fsa, 'aux_labels'):
+        return k2.intersect(a_fsa, b_fsa)
+
+    if not hasattr(b_fsa, 'aux_labels'):
+        return k2.intersect(a_fsa, b_fsa)
+
+    a_fsa = a_fsa.invert()
+
+    a_fsa = k2.arc_sort(a_fsa)
+    b_fsa = k2.arc_sort(b_fsa)
+
+    need_arc_map = True
+    treat_epsilons_specially = True
+    ragged_arc, a_arc_map, b_arc_map = _k2.intersect(
+        a_fsa.arcs, a_fsa.properties, b_fsa.arcs, b_fsa.properties,
+        treat_epsilons_specially, need_arc_map)
+
+    out_fsa = k2.Fsa(ragged_arc)
+    out_fsa.labels = k2.index_attr(a_fsa.aux_labels, a_arc_map)
+    out_fsa.aux_labels = k2.index_attr(b_fsa.aux_labels, b_arc_map)
+
+    for name, a_value in a_fsa.named_tensor_attr():
+        if hasattr(b_fsa, name):
+            # Both a_fsa and b_fsa have this attribute.
+            # We only support attributes with dtype `torch.float32`.
+            # Other kinds of attributes are discarded.
+            if a_value.dtype != torch.float32:
+                continue
+            b_value = getattr(b_fsa, name)
+            assert b_value.dtype == torch.float32
+
+            value = k2.index_select(a_value, a_arc_map) + k2.index_select(
+                b_value, b_arc_map)
+            setattr(out_fsa, name, value)
+        else:
+            # only a_fsa has this attribute, copy it via arc_map
+            value = k2.index_attr(a_value, a_arc_map)
+            setattr(out_fsa, name, value)
+
+    # now copy tensor attributes that are in b_fsa but are not in a_fsa
+    for name, b_value in b_fsa.named_tensor_attr():
+        if not hasattr(out_fsa, name):
+            value = k2.index_attr(b_value, b_arc_map)
+            setattr(out_fsa, name, value)
+
+    for name, a_value in a_fsa.named_non_tensor_attr():
+        if name == 'symbols':
+            continue
+
+        if name == 'aux_symbols':
+            setattr(out_fsa, 'symbols', a_value)
+        else:
+            setattr(out_fsa, name, a_value)
+
+    for name, b_value in b_fsa.named_non_tensor_attr():
+        if not hasattr(out_fsa, name):
+            setattr(out_fsa, name, b_value)
+
+    return out_fsa
+
+
+class CtcTrainingGraphCompiler:
```

Comment:
BTW I think you can just remove the previous graph compiler code as it was essentially a half-baked CTC, I don’t think we’ll need it going forward?

Reply:
Removed.